### PR TITLE
CO-1882: Group management updates

### PR DESCRIPTION
**Pull request by spartanicus**
This PR holds early improvements in group management, specifically CO-1916 (Break Group Management Page Into Component Tabs) and CO-1925 (Default Group Filtering Should Exclude Automatic Groups). The latter includes a simple API for managing session variables via ajax calls (SessionXhrController.php).

### CODE REVIEW

**Comment by spartanicus** [comment]
This PR should be reworked slightly to adopt the newly merged Bootstrap library (with MDL now deprecated). Also - the console.log() items should be removed from the code.

**Comment by spartanicus** [comment]
Refactoring now done to adopt Bootstrap for this PR (and console.log() items have been removed).

**Comment by spartanicus** [comment]
The feature-co1882 branch has been rebased against develop.

**Comment by boshrin** [comment]
Next steps are to performance test people picker autocomplete box (can use QA), then ask for feedback from dev group.

**Comment by spartanicus** [comment]
This has been rebased against the latest develop.

**Review by boshrin — COMMENTED** [review]
I see both `$nest_contains` and `$select_contains` defined but not used anywhere?

**Comment by spartanicus** [comment]
This has been rebased against the latest develop

**Comment by spartanicus** [comment]
Rebased against the latest develop.

File: app/Controller/CoGroupMembersController.php [hunk 1]
```diff
@@ -759,13 +830,53 @@ public function updateGroup() {
...
+      //$rid = $this->Session->read('Auth.User.co_person_id'); // used in setMembership()
+      try {
+        if($this->CoGroupMember->isMember($gid,$pid)) {
+          // CoPerson is already a member
+          $this->Flash->set(_txt('er.grm.already',array($plb,$gnm)), array('key' => 'error'));
+        } else {
+          // Add CoPerson to the group
+          $this->CoGroupMember->addByGroupName($pid,$gnm);
+          // XXX This is likely the better choice - needs RoleComponent (the first parameter)
+          // $this->CoGroupMember->setMembership(null,$gid,$pid,true,false,$rid);
```
Comment (boshrin) [inline app/Controller/CoGroupMembersController.php:866]: Should this be addressed or removed?

File: app/Controller/CoGroupsController.php [hunk 2]
```diff
@@ -74,7 +74,7 @@ class CoGroupsController extends StandardController {
    */
 
   function beforeFilter() {
-    $this->redirectTab = 'group';
+    $this->redirectTab = 'group'; // XXX legacy? remove?
```
Comment (boshrin) [inline app/Controller/CoGroupsController.php:77]: Yes this is legacy and can be removed (and therefore the whole `beforeFilter()` override).

File: app/Controller/CoGroupsController.php [hunk 3]
```diff
@@ -227,24 +227,40 @@ function checkWriteDependencies($reqdata, $curdata = null) {
...
-          $this->Flash->set(_txt('er.gr.init'), array('key' => 'information'));
-          return false;
+    if(!$this->request->is('restful')) {
+
+      if($this->action == 'add') {
+        $cos = $this->Session->read('Auth.User.cos');
+        $this->redirectTarget = array('action' => 'edit', $this->CoGroup->id);
+
+        // Member of current CO? (Platform admin wouldn't be)
+        if (isset($cos) && isset($cos[$this->cur_co['Co']['name']]['co_person_id'])) {
```
Comment (boshrin) [inline app/Controller/CoGroupsController.php:237]: A platform admin _could_ be a member of the current CO, does that matter?

Also, if you test `isset($foo['bar'])` you don't need to also test `isset($foo)`.

File: app/Controller/CoGroupsController.php [hunk 4]
```diff
@@ -227,24 +227,40 @@ function checkWriteDependencies($reqdata, $curdata = null) {
...
         }
       }
+
+      if($this->action == 'edit') {
+        // return to the index view with default filters in place after edit
+        $this->redirectTarget = array(
+          'controller' => 'co_groups',
+          'action' => 'index',
+          'co' => filter_var($this->cur_co['Co']['id'],FILTER_SANITIZE_SPECIAL_CHARS),
+          $this->CoGroup->id,
```
Comment (boshrin) [inline app/Controller/CoGroupsController.php:258]: This is confusing and non-standard... why does and `index` view take a record id?

File: app/Controller/CoGroupsController.php [hunk 5]
```diff
@@ -465,6 +555,19 @@ function isAuthorized() {
     // View an existing Group?
     $p['view'] = ($roles['cmadmin'] || $roles['coadmin'] || $managed);
     
+    // Viewing members, email lists, and nested groups is available to all group members.
+    // Access to specific actions is controlled in the view.
+    $p['members'] = $roles['cmadmin'] || $roles['coadmin'] || $managed || $member;
+    $p['email_lists'] = $roles['cmadmin'] || $roles['coadmin'] || $managed || $member;
+    $p['nest'] =  $roles['cmadmin'] || $roles['coadmin'] || $managed || $member;
```
Comment (boshrin) [inline app/Controller/CoGroupsController.php:562]: I'm not sure `|| $member` is appropriate here. The only permission that was previously intended was for a member to see their own membership (ie: the `co_group_member`) view, and maybe that wasn't even fully thought out.

File: app/Controller/CoGroupsController.php [hunk 6]
```diff
@@ -681,6 +792,64 @@ public function paginationConditions() {
       $pagcond['conditions']['CoGroup.group_type'] = $searchterm;
     }
 
+    // Exclude admin groups
+    if(!empty($this->params['named']['search.noadmin'])) {
+      $pagcond['conditions']['CoGroup.group_type <>'] = 'A';
```
Comment (boshrin) [inline app/Controller/CoGroupsController.php:797]: Use enums, not values.

```suggestion
      $pagcond['conditions']['CoGroup.group_type <>'] = GroupEnum::Admins;
```

File: app/Controller/CoGroupsController.php [hunk 7]
```diff
@@ -681,6 +792,64 @@ public function paginationConditions() {
...
+    // Filter by membership and ownership
+    // First get the CoPersonID to work against either the current user, or the user being acted on in select mode
+    $coPersonId = $this->Session->read('Auth.User.co_person_id');
+    if($this->action == 'select') {
+      $coPersonId = filter_var($this->request->params['named']['copersonid'], FILTER_SANITIZE_SPECIAL_CHARS);
+    }
+
+    // If both member and owner are selected, use a single join with "OR"
+    if(!empty($this->params['named']['search.member']) && !empty($this->params['named']['search.owner'])) {
+      $pagcond['joins'][] = array(
```
Comment (boshrin) [inline app/Controller/CoGroupsController.php:809]: Joins need to be tested against ChangelogBehavior. The easiest way to do this is to make changes in a record of the target table and then make sure you pull the newer values, not the older ones.

File: app/Lib/enum.php [hunk 8]
```diff
@@ -499,6 +499,7 @@ class OrgIdentityStatusEnum
 class PeoplePickerModeEnum
 {
   const Sponsor   = 'S';
+  const CoPerson  = 'CP';
```
Comment (boshrin) [inline app/Lib/enum.php:502]: Is this the right name? People Pickers are always picking CO People, the mode is intended to convey the context. So it should probably be something like `GroupMember`, or if there's nothing specific to groups then maybe `All` or `Active`.

File: app/Lib/lang.php [hunk 9]
```diff
@@ -704,6 +704,12 @@
                                  NameEnum::Official => 'Official',
                                  NameEnum::Preferred => 'Preferred'),
   
+  // Nested Group member filters
+  'en.nested.filters' => array(
+    'e'      => 'Exclude',
```
Comment (boshrin) [inline app/Lib/lang.php:709]: A proper enum should be defined in `enum.php`.

File: app/View/CoGroupMembers/select.ctp [hunk 10]
```diff
@@ -79,98 +181,231 @@
...
+                  } else {
+                    print _txt('fd.group.own.only');
+                  }
+                } elseif($isMember) {
+                  print _txt('fd.group.mem');
+                }
+              }
+
+              // Warn if membership is invalid
+              // XXX Use badge after introduced by PR-178 / CO-2054
```
Comment (boshrin) [inline app/View/CoGroupMembers/select.ctp:301]: PR-178 is merged?

File: app/Lib/enum.php (hunk 8 above)
Comment (spartanicus) [reply app/Lib/enum.php:502]: Good point - I think this mode can be used anywhere to grab a person; perhaps "All" is correct - I don't believe we're limiting the Group Member picker to Active only.

File: app/Controller/CoGroupMembersController.php (hunk 1 above)
Comment (spartanicus) [reply app/Controller/CoGroupMembersController.php:833]: It should be removed.

File: app/Controller/CoGroupsController.php (hunk 2 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php:77]: Excellent - will do :-)

File: app/Controller/CoGroupsController.php (hunk 3 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php:237]: Shouldn't matter - agree that the comment is inaccurate. Will fix up comment and the test!

File: app/Controller/CoGroupsController.php (hunk 5 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php:562]: Yeah - this is worth a discussion. We can definitely scale this back - though it does beg the question of what a member sees when they visit a group they are in.

File: app/Lib/lang.php (hunk 9 above)
Comment (spartanicus) [reply app/Lib/lang.php]: I've updated the values, but the issues remains. I'll add these to enum.php as well.

File: app/View/CoGroupMembers/select.ctp (hunk 10 above)
Comment (spartanicus) [reply app/View/CoGroupMembers/select.ctp:301]: Yes! I'll get that in.

File: app/Controller/CoGroupsController.php (hunk 4 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php:258]: The record id was accidentally left in (bad copy paste, I suspect). Removed!

File: app/Controller/CoGroupsController.php (hunk 6 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php]: Fix committed

File: app/Lib/enum.php (hunk 8 above)
Comment (spartanicus) [reply app/Lib/enum.php:502]: Went with `const All       = 'AL';` for now

File: app/View/CoGroupMembers/select.ctp (hunk 10 above)
Comment (spartanicus) [reply app/View/CoGroupMembers/select.ctp:301]: Yes! I'll look at this.

File: app/Lib/lang.php (hunk 9 above)
Comment (spartanicus) [reply app/Lib/lang.php]: Added - we now have NestedEnum of which both lang.php and the CoGroupMembersController make use.

File: app/Controller/CoGroupMembersController.php (hunk 1 above)
Comment (spartanicus) [reply app/Controller/CoGroupMembersController.php]: Update committed as of ee3fa84

File: app/Controller/CoGroupsController.php (hunk 2 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php]: Update committed as of ee3fa84

File: app/Controller/CoGroupsController.php (hunk 3 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php]: Update committed as of ee3fa84

File: app/Controller/CoGroupsController.php (hunk 4 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php]: Update committed as of ee3fa84

File: app/Lib/enum.php (hunk 8 above)
Comment (spartanicus) [reply app/Lib/enum.php]: Update committed as of ee3fa84

File: app/Lib/lang.php (hunk 9 above)
Comment (spartanicus) [reply app/Lib/lang.php]: Update committed as of ee3fa84

**Comment by spartanicus** [comment]
First set of updates is complete, leaving three comments above still pending. I'm now looking at the joins (for changelog behavior) and the possible badging for invalid memberships. 

It's worth discussing what a regular member sees in a group (if anything) before I change the access to the group tabs for regular members. In develop, a regular member can see the Group in "view" mode - which provides access to the group properties, identifiers, nested groups, email lists, and the full members listing. This PR retains that behavior except for hiding the identifiers (which regular members don't need to see). 

The view of a Group is easily locked down from members, but how far do we lock it down? Do members see other members? Can members only see the group name in the list of their groups but not access the group view itself? Is that a Group RFE for setting on a case-by-case basis or CO setting? The current behavior is the same as develop, but we can certainly change it.

File: app/Controller/CoGroupsController.php (hunk 5 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php:546]: See [comment](https://github.com/Internet2/comanage-registry/pull/185#issuecomment-910264336) down below

File: app/Controller/CoGroupsController.php (hunk 7 above)
Comment (spartanicus) [reply app/Controller/CoGroupsController.php:793]: This appears to be working; changes to member or owner status are immediately reflected in the pagination of the group (and group members) listing. (Tested: changes to member/owner status by an admin. The member immediately saw the change on reload/clicking a tab. Also tested updates to nested groups.)

The "Nested Groups" and "Email Lists" tabs both display the changelog at the bottom of their views, but these only reflect the Group properties changes. I suspect these should be removed so the context is clear. Do we need changelog behavior for these? The "Provisioned Services" has a changelog display too.

**Comment by boshrin** [comment]
Based on today's dev call

1. "My Groups" and "My Memberships" should not show up in the menu for CMP admins who are not members of the CO.
2. "Groups" and "All Groups" need some clarification. We could have (eg) "Regular Groups", "System Groups", and "All Groups". Or "All Groups (unfiltered)". Or "Groups (default filter)".
3. "Provision" button should not render for COU admins, though they should be able to see the tab.

**Comment by spartanicus** [comment]
"Regular Groups", "System Groups", and "All Groups" makes the most sense to me for item 1 - I think that distinguishes the difference between them fairly well and provides a nice shortcut to the "auto" groups. We'll start with that.

**Comment by spartanicus** [comment]
This PR has been rebased against the latest develop.

**Comment by spartanicus** [comment]
Notes 1 & 2 above are now fixed: "My Groups" and  "My Memberships" only appear for an active CoPerson in a CO, and the Group main menu links have been changed. 

This leaves three (known) items: read-only members list pagination (most important), restricting the "provision" button, and badges for invalid memberships.

**Comment by spartanicus** [comment]
This has been rebased against the latest develop.

**Comment by spartanicus** [comment]
Rebased again along with a fix that hides the Group "Provision" buttons from COU Admins.

**Pull request merged**